Run e2e tests on alfajores

piersy opened the pull request:
Modifies our e2e tests to be able to run on alfajores if the environment variable `NETWORK` is set to 'alfajores'.
Also adds a CI step to run them against alfajores.

Note that the `test_fee_currency...` tests are skipped because they require the ability to deploy new fee currencies, which we probably don't want to be doing on a live network as it could cause issues for people, and we also don't want to expose the credentials to do so.

piersy committed:
Update viem

piersy committed:
Improve error handling and logging

piersy committed:
Make test fee currency addresses more searchable

piersy committed:
Add support for running against alfajores

piersy committed:
Wait for first failed transaction

As opposed to sending it async, this ensures that we don't end up with
an invalid nonce on the next transaction.

piersy committed:
Ensure that tail processes are killed

piersy committed:
Add ci step to run e2e tests on alfajores

piersy committed:
Update tests to work on alfajores

piersy committed:
Fix just enough test

piersy commented on e2e_test/js-tests/test_viem_tx.mjs:
```diff
@@ -271,27 +309,38 @@ describe("viem send tx", () => {
-		// currency into account everywhere.
-		const maxFeePerGas = block.baseFeePerGas / 2n;
+		const maxFeePerGas = rate.toFeeCurrency(block.baseFeePerGas)+2n;
 		const request = await walletClient.prepareTransactionRequest({
 			account,
 			to: "0x00000000000000000000000000000000DeaDBeef",
 			value: 2,
-			gas: 90000,
-			feeCurrency: process.env.FEE_CURRENCY2,
-			maxFeePerGas,
-			maxPriorityFeePerGas: 1n,
+			gas: 171000,
+			feeCurrency: process.env.FEE_CURRENCY,
+			maxFeePerGas: maxFeePerGas,
+			maxPriorityFeePerGas: 2n,
```
Hey @karlb, I'm thinking probably my modifications here don't make sense because it seems that it is important for this test that the fee per gas for the fee currency than the native currency, and I've modified it to use USDC which is the other way around.

ezdac replied:
It seems that the main point of this is that the fee should be too low for native currency, but if the node converts it then it is sufficient. But for USDC it should actually be fine, since the exchange rate results in requiring a lower value for the base-fee compared to when it is denominated in Celo.

piersy replied:
I will leave as is, with a comment that we may want to change this if the exchange rate changes significantly.

piersy committed:
Add debug

piersy committed:
Re-instate previous log capture kill approach

piersy committed:
Fix script dir location

piersy committed:
Ensure that send_tx script exits once done

Without this it can hang for about 12 minutes!

piersy committed:
Ensure that we match system addresses irrespective of casing

ezdac commented on e2e_test/js-tests/test_viem_tx.mjs:
```diff
@@ -124,15 +141,16 @@ describe("viem send tx", () => {
 
-	it("send fee currency tx and check receipt", async () => {
+	it("send fee currency tx with explicit gas fields and check receipt", async () => {
+		const [maxFeePerGas, tip] = await getGasFees(publicClient, 2n);
```
Ideally this would also take a fee-currency parameter and convert the gas-fees into a fee-currency denominated one. The method e.g. will not work in cReal, because the multiplier required is higher than 2

piersy replied:
Thanks, will update!

ezdac commented on e2e_test/js-tests/test_viem_tx.mjs:
```diff
@@ -63,7 +79,8 @@ const testNonceBump = async (
 			to: "0x00000000000000000000000000000000DeaDBeef",
 			value: 3,
-			gas: 90000,
+			// dynaimcally retrieve intrinsic gas from feeCurrency directory "feeCurrencyConfig"
```
Do you still plan to implement this for this PR? If not, I would remove the line or add a `TODO:`, because it doesn't reflect what's happening ;)

piersy replied:
Ok removed and I added a ticket for it - https://github.com/celo-org/op-geth/issues/260

ezdac reviewed (commented):
I didn't manage to look at everything in detail yet.
But looks very good so far - this feature is much appreciated.

For future changes: I think we should add more fee-currencies registered in the directory and test this for multiple currencies.

karlb commented on e2e_test/debug-fee-currency/lib.sh:
```diff
@@ -13,16 +13,16 @@ set -xeo pipefail
 function deploy_fee_currency() {
-	(
-		local fee_currency=$(
+	(set -e;
+		local fee_currency=$(set -e;
```
We have no common bash style, but my though is we should have -e set everywhere, so this is unnecessary noise. But since we don't have much bash library code it's fine either way.

piersy replied:
This was a mistake on my part, I think they can be removed.

karlb commented on e2e_test/js-tests/package.json:
Doesn't npm always use two spaces for indentation? Why this formatting change here and in `package-lock.json`?

piersy replied:
Thanks missed that, updated my json formatter. Also I noticed that npm follows the indentation in package.json for the package-lock.json, so that is also looking neater now.

karlb commented on e2e_test/run_all_tests.sh:
```diff
@@ -32,6 +34,14 @@ cast send --async --json --private-key "$ACC_PRIVKEY" "$TOKEN_ADDR" 'transfer(ad
+	if [[ -n $NETWORK ]]; then
+		case $f in
+		  # Skip tests that require a local network.
+		  test_fee_currency_fails_on_credit.sh|test_fee_currency_fails_on_debit.sh|test_fee_currency_fails_intrinsic.sh)
```
This is correct, but the combination of bash's case syntax and the regex made me look for the left parenthesis far too long. (no change need, just leaving random thoughts)

piersy replied:
It's a bit weird isn't it!

karlb commented on e2e_test/js-tests/test_viem_tx.mjs:
```diff
@@ -271,27 +309,38 @@ describe("viem send tx", () => {
 	}).timeout(10_000);
 
-	it("send fee currency tx with just high enough gas price", async () => {
+	it("send fee currency tx with just high enough gas price mine", async () => {
```
```suggestion
	it("send fee currency tx with just high enough gas price", async () => {
```

piersy replied:
Thanks, removed.

karlb commented on e2e_test/js-tests/test_viem_tx.mjs:
```diff
@@ -271,27 +309,38 @@ describe("viem send tx", () => {
-		// exchange rate of 2, so the test will still check that we take the fee
-		// currency into account everywhere.
-		const maxFeePerGas = block.baseFeePerGas / 2n;
+		const maxFeePerGas = rate.toFeeCurrency(block.baseFeePerGas)+2n;
```
In case of increasing base fees, this might not be enough to get the tx into a block.

piersy replied:
What do you suggest? Right now this seems to be working reliably, but could get flaky pretty easily.

karlb replied:
We could add a small amount (maybe 10%, but must be lower than the exchange rate between fee currency and native token) to reduce the likelihood of having to wait for a low enough base fee.

piersy replied:
Hey @karlb, I've updated the test to address your suggestion.

karlb commented on e2e_test/js-tests/test_viem_tx.mjs:
```diff
@@ -271,27 +309,38 @@ describe("viem send tx", () => {
 		}
 	}).timeout(10_000);
 
-	it("send fee currency tx with just high enough gas price", async () => {
+	it("send fee currency tx with just high enough gas price mine", async () => {
+		const rate = await getRate(process.env.FEE_CURRENCY);
 		const block = await publicClient.getBlock({});
-		// Actually, the base fee will be a bit lower next block, since our blocks
-		// are always mostly empty. But the difference will be much less than the
-		// exchange rate of 2, so the test will still check that we take the fee
-		// currency into account everywhere.
-		const maxFeePerGas = block.baseFeePerGas / 2n;
+		const maxFeePerGas = rate.toFeeCurrency(block.baseFeePerGas)+2n;
 		const request = await walletClient.prepareTransactionRequest({
 			account,
 			to: "0x00000000000000000000000000000000DeaDBeef",
 			value: 2,
-			gas: 90000,
-			feeCurrency: process.env.FEE_CURRENCY2,
-			maxFeePerGas,
-			maxPriorityFeePerGas: 1n,
+			gas: 171000,
+			feeCurrency: process.env.FEE_CURRENCY,
+			maxFeePerGas: maxFeePerGas,
+			maxPriorityFeePerGas: 2n,
 		});
 		const signature = await walletClient.signTransaction(request);
 		const hash = await walletClient.sendRawTransaction({
 			serializedTransaction: signature,
 		});
 		const receipt = await publicClient.waitForTransactionReceipt({ hash });
 		assert.equal(receipt.status, "success", "receipt status 'failure'");
-	}).timeout(5_000);
+	}).timeout(10_000);
```
Is this necessary? With 1 second block times, 5 seconds should be plenty.

I can understand the increase for testing against a live network, but when iterating on local changes, quick failure is a nice feature. Most failures during development are not of the "tx does not get included" kind though.

piersy replied:
So this is the running time on alfajores.

```
    ✔ send fee currency tx with just high enough gas price (4998ms)
```

I generally like to have a reasonable sized buffer for timeouts to avoid failures based on minor resource availability fluctuations. So I'd argue it is necessary.

 If you feel strongly about this we could also add custom timeouts per network.

karlb replied:
In that case, increasing the timeout makes sense. I expected the test to be much faster on alfajores.

karlb commented on e2e_test/test_fee_currency_fails_intrinsic.sh:
```diff
@@ -7,6 +7,7 @@ source debug-fee-currency/lib.sh
 
 # Expect that the creditGasFees failed and is logged by geth
 tail -F -n 0 geth.log >debug-fee-currency/geth.intrinsic.log & # start log capture
+# trap 'kill %%' EXIT # kill bg job at exit
```
If this works, we should use it. If not, I would like to know why are either add that as a comment or remove this line.

piersy replied:
I was fiddling around with this for a while couldn't understand why the test was hanging. In fact the hang was caused by `send_tx.mjs` not exiting (fixed by calling `process.exit(0)`). But also if the test was interrupted you could end up with multiple tails running in the background, which is what brought me to this.

I've re-instated it and removed the other kill command.

karlb reviewed (commented):
Good feature and a bunch of small improvements on top of that!

piersy committed:
Fix typo

piersy committed:
Add comment to just high enough test

piersy committed:
Ensure we get gas fees in fee currency

piersy committed:
Remove stray comment

piersy committed:
Remove un-needed bash option setting

piersy committed:
Correct indentation

piersy committed:
Kill bg jobs with trap

This ensures the jobs get killed however the script exits.

piersy committed:
Remove debug logs

piersy committed:
Add 10% buffer to base fee in just enough test

Not that this test is running on live networks we need to account for
the case where the base fee could increase, we do this by adding a 10%
buffer.

Also ensures that we use FEE_CURRENCY2 when running local tests.

And finally adds a check to fail the test if the value of the fee
currency is not sufficiently greater than the value of celo.
